compiler/rustc_mir/src/transform/remove_zsts.rs

```diff
@@ -0,0 +1,36 @@
+//! Removes assignments to ZST places.
+
+use crate::transform::MirPass;
+use rustc_middle::mir::{Body, StatementKind};
+use rustc_middle::ty::TyCtxt;
+
+pub struct RemoveZsts;
+
+impl<'tcx> MirPass<'tcx> for RemoveZsts {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let param_env = tcx.param_env(body.source.def_id());
+        let (basic_blocks, local_decls) = body.basic_blocks_and_local_decls_mut();
+        for block in basic_blocks.iter_mut() {
+            for statement in block.statements.iter_mut() {
+                match statement.kind {
+                    StatementKind::Assign(box (place, _)) => {
+                        let place_ty = place.ty(local_decls, tcx).ty;
+                        if let Ok(layout) = tcx.layout_of(param_env.and(place_ty)) {
```

Comment:
Maybe a fast path for known ZSTs (well, let's start with just `()`) could reduce the number of query calls?